rqd/rqd/rqcore.py

```diff
@@ -1080,3 +1060,51 @@ def sendStatusReport(self):
     def isWaitingForIdle(self):
         """Returns whether the host is waiting until idle to take some action."""
         return self.__whenIdle
+
+    def sendFrameCompleteReport(self, runningFrame):
+        """Send a frameCompleteReport to Cuebot"""
+        if not runningFrame.completeReportSent:
+            report = rqd.compiled_proto.report_pb2.FrameCompleteReport()
+            # pylint: disable=no-member
+            report.host.CopyFrom(self.machine.getHostInfo())
+            report.frame.CopyFrom(runningFrame.runningFrameInfo())
+            # pylint: enable=no-member
+
+            if runningFrame.exitStatus is None:
+                report.exit_status = 1
+            else:
+                report.exit_status = runningFrame.exitStatus
+
+            report.exit_signal = runningFrame.exitSignal
+            report.run_time = int(runningFrame.runTime)
+
+            # If nimby is active, then frame must have been killed by nimby
+            # Set the exitSignal to indicate this event
+            if self.nimby.locked and not runningFrame.ignoreNimby:
+                report.exit_status = rqd.rqconstants.EXITSTATUS_FOR_NIMBY_KILL
+
+            self.network.reportRunningFrameCompletion(report)
+            runningFrame.completeReportSent = True
+
+    def sanitizeFrames(self):
+        """
+        Iterate over the cache and update the status of frames that might have
+        completed but never reported back to cuebot.
+        """
+        for frameId, runningFrame in self.__cache.items():
+            # If the frame was marked as completed (exitStatus) and a report has not been sent
+            # try to file the report again
+            if runningFrame.exitStatus is not None and not runningFrame.completeReportSent:
+                try:
+                    self.sendFrameCompleteReport(runningFrame)
+                    self.deleteFrame(frameId)
+                    self.info("Successfully deleted frame from cache for %s/%s (%s)" % (
+                                  runningFrame.runFrame.job_name,
+                                  runningFrame.runFrame.frame_name,
+                                  frameId))
+                # pylint: disable=broad-except
+                except Exception:
```

Comment:
log.exception("Failed to sanitize frame %s/%s",
                                  runningFrame.runFrame.job_name,
                                  runningFrame.runFrame.frame_name)